File: scala/fn_currying.scala


// context-bound - same as
// def mul[T](x:T, y:T)(implicit n:Numeric[T]) = n.times(x,y)
def mul[T:Numeric](x:T, y:T) = implicitly[Numeric[T]].times(x,y)

// partial application
def mul3[T:Numeric](x:T) = mul[T](3,x)
//def mul_curried[T](x:Numeric[T])(y:Numeric[T]) = mul(x,y)

println(mul(3,5))
println(mul3(5))

File: old/haskell/fn_currying.hs
-- named after haskell curry
-- refer https://www.haskell.org/haskellwiki/Currying

-- all functions take one parameter!

-- f x y = bar === f x = \y -> bar === f = \x -> \y -> bar

mul x y = x * y

-- mul taking one param gives a function
-- mul 3 is a partial application of mul
mul3 = mul 3


-- in prelude you can type mul_uncurried = uncurry mul
mul_uncurried (x,y) = mul x y

mul_curried x y = mul_uncurried (x,y)

main = do
		-- print mul3
	    print . mul3 $ 5
	    print $ mul_uncurried (3,5)
	    print $ mul_curried 3 5

